sycl/source/detail/persistent_device_code_cache.cpp

```diff
@@ -0,0 +1,393 @@
+    RT::PiProgram &NativePrg) {
+
+  if (!isEnabled())
+    return {};
+
+  // Only SPIRV images are cached
+  if (Img.getFormat() != PI_DEVICE_BINARY_TYPE_SPIRV)
+    return {};
+
+  std::string Path =
+      getCacheItemPath(Device, Img, SpecConsts, BuildOptionsString);
+
+  if (!isPathPresent(Path))
+    return {};
+
+  int i = 0;
+
+  // If cache directory is locked ignore cache
+  if (LockCacheItem::isLocked(Path))
+    return {};
+
+  std::string FileName{Path + "/" + std::to_string(i)};
+  while (isPathPresent(FileName + ".bin") || isPathPresent(FileName + ".src")) {
+    if (isCacheItemSrcEqual(FileName + ".src", Device, Img, SpecConsts,
+                            BuildOptionsString)) {
+      try {
+        return readBinaryDataFromFile(FileName + ".bin");
+      } catch (...) {
+        // If read was unsuccessfull try the next item
+      }
+    }
+    FileName = Path + "/" + std::to_string(++i);
+  }
+  return {};
+}
+
+/* Returns string value which can be used to identify different device
+ */
+std::string PersistentDeviceCodeCache::getDeviceIDString(const device &Device) {
+  return Device.get_platform().get_info<sycl::info::platform::name>() + "/" +
+         Device.get_info<sycl::info::device::name>() + "/" +
+         Device.get_info<sycl::info::device::version>() + "/" +
+         Device.get_info<sycl::info::device::driver_version>();
+}
+
+/* Write built binary to persistent cache
+ * Format: numImages, 1stImageSize, Image[, NthImageSize, NthImage...]
+ * Return on first unsuccessfull file operation
+ */
+void PersistentDeviceCodeCache::writeBinaryDataToFile(
+    const std::string &FileName, const std::vector<std::vector<char>> &Data) {
+  std::ofstream FileStream{FileName, std::ios::binary};
+  if (FileStream.fail())
+    return;
+
+  size_t Size = Data.size();
+  FileStream.write((char *)&Size, sizeof(Size));
+  if (FileStream.fail())
+    return;
+
+  for (size_t i = 0; i < Data.size(); ++i) {
+    Size = Data[i].size();
+    FileStream.write((char *)&Size, sizeof(Size));
+    if (FileStream.fail())
+      return;
+    FileStream.write(Data[i].data(), Size);
+    if (FileStream.fail())
+      return;
+  }
+  FileStream.close();
+}
+
+/* Read built binary to persistent cache
+ * Format: numImages, 1stImageSize, Image[, NthImageSize, NthImage...]
+ */
+std::vector<std::vector<char>>
+PersistentDeviceCodeCache::readBinaryDataFromFile(const std::string &FileName) {
+  std::ifstream FileStream{FileName, std::ios::binary};
+  if (FileStream.fail())
+    return {};
+  size_t ImgNum, ImgSize;
+  FileStream.read((char *)&ImgNum, sizeof(ImgNum));
+  if (FileStream.fail())
+    return {};
+
+  std::vector<std::vector<char>> Res(ImgNum);
+  for (size_t i = 0; i < ImgNum; ++i) {
+    FileStream.read((char *)&ImgSize, sizeof(ImgSize));
+    if (FileStream.fail())
+      return {};
+
+    std::vector<char> ImgData(ImgSize);
+    FileStream.read(ImgData.data(), ImgSize);
+    if (FileStream.fail())
```

Comment:
Nit: here and in many other places: fail status can be checked once before returning, badbit is sticky, I believe (which means once operation fails and sets it, subsequent operations won't clear it). That would give better code readability, I believe.